backend-ktor/src/main/kotlin/me/zwsmith/plugins/database/RecipeService.kt

```diff
@@ -0,0 +1,135 @@
+        }
+    }
+    
+    suspend fun createIngredient(name: String): Ingredient = dbQuery {
+        IngredientDao.new {
+            this.name = name
+        }.toIngredient()
+    }
+    
+    suspend fun findIngredientById(id: Int) = dbQuery {
+        IngredientDao.findById(id)?.toIngredient()
+    }
+    
+    suspend fun updateIngredient(updateData: Ingredient) = dbQuery {
+        IngredientDao.findById(updateData.id)?.apply {
+            name = updateData.name
+        }?.toIngredient()
+    }
+    
+    suspend fun deleteIngredient(id: Int) = dbQuery {
+        val ingredient = IngredientDao.findById(id)
+        if(ingredient == null) {
+            throw IllegalArgumentException("No Ingredient found with ID $id")
+        } else {
+            ingredient.delete()
+        }
+    }
+    
+    suspend fun createRecipe(
+        name: String,
+        description: String,
+        ingredientInputs: List<QuantifiedIngredientInput>
+    ): Recipe {
+        return dbQuery {
+            val recipe = RecipeDao.new {
+                this.name = name
+                this.description = description
+            }
+            
+            // Associate QuantifiedIngredients with the new Recipe
+            ingredientInputs.forEach { input ->
+                val ingredient = IngredientDao
+                    .find { Ingredients.name eq input.ingredient.name }
+                    .singleOrNull()
+                    ?: IngredientDao.new {
+                        this.name = input.ingredient.name
+                    }
+                QuantifiedIngredientDao.new {
+                    this.amount = input.amount
+                    this.unit = input.unit
+                    this.recipe = recipe
+                    this.ingredient = ingredient
+                }
+            }
+            
+            recipe.toRecipe()
+        }
+    }
+    
+    suspend fun findRecipeById(id: Int): Recipe = dbQuery {
+        RecipeDao.findById(id)
+            ?.toRecipe()
+            ?: throw IllegalArgumentException("Recipe with id $id does not exist")
+    }
+    
+    suspend fun findRecipes(): List<Recipe> = dbQuery {
+        RecipeDao.all().map { it.toRecipe() }
+    }
+    
+    suspend fun updateRecipe(
+        recipeId: Int,
+        name: String? = null,
+        description: String? = null,
+        ingredients: List<QuantifiedIngredient>? = null
+    ): Recipe = dbQuery {
+        var recipe = RecipeDao.findById(recipeId)
+            ?: throw IllegalArgumentException("Recipe with id $recipeId not found")
+        if (name != null) {
+            recipe.name = name
+        }
+        if (description != null) {
+            recipe.description = description
+        }
+        if(ingredients != null) {
+            updateRecipeIngredients(recipeId, ingredients)
+        }
+        recipe.toRecipe()
+    }
+    
+    suspend fun updateRecipeIngredients(recipeId: Int, ingredients: List<QuantifiedIngredient>) = dbQuery {
```

Comment:
This name should be `replaceRecipeIngredients`

I would expect `updateRecipeIngredients` to not delete existing ingredients entirely, but rather update quantities of existing ingredients, and insert ingredients not already present.

Reply:
yeah that's fine, I just think you should update the name to better reflect what's actually happening.

If you need an update ingredients later, you can add it later

Reply:
I was actually able to figure out how to update existing relations and add new ones in the one transaction. No need to delete

Reply:
Not sure if the error handling is optimal though. I wanted it to throw so that it would rollback the transaction if any of the new ingredient relationships passed don't exist

Reply:
Invoke [`rollback`](https://jetbrains.github.io/Exposed/api/exposed-core/org.jetbrains.exposed.sql/-transaction/rollback.html) to abort the transaction